lib/posix-process/exit.c

```diff
@@ -0,0 +1,318 @@
+	pprocess_release_pthread(pthread);
+	if (thread != uk_thread_current())
+		uk_sched_thread_terminate(thread);
+}
+
+void pprocess_exit(struct posix_process *pprocess,
+		   enum posix_process_state state,
+		   int exit_status)
+{
+	struct posix_process_exit_event_data event_data;
+	struct posix_process *parent_process;
+	struct posix_thread *pt, *ptn;
+	__bool nowait = false;
+	int ret;
+
+	UK_ASSERT(state == POSIX_PROCESS_EXITED ||
+		  state == POSIX_PROCESS_KILLED);
+
+	UK_ASSERT(pprocess);
+	UK_ASSERT(pprocess->state == POSIX_PROCESS_RUNNING);
+
+	uk_pr_debug("pid %d: exit process\n", pprocess->pid);
+
+	/* May be NULL if init terminated early */
+	parent_process = pprocess->parent;
+
+	/* As this will be signaling the parent on behalf of the current pthread,
+	 * we must do this before we terminate the pthreads, in case we are
+	 * called from _exit(), exit_group() (i.e. operate on current).
+	 */
+	if (parent_process) {
+#if CONFIG_LIBPOSIX_PROCESS_SIGNAL
+		ret = signal_exit(parent_process);
+		if (ret > 0) {
+			/* From exit(2): "If the parent has set SA_NOCLDWAIT, or has
+			 * set the SIGCHLD handler to SIG_IGN, the status is discarded
+			 * and the child dies immediately."
+			 */
+			uk_pr_info("Parent ignores SIGHLD, terminating\n");
+			nowait = true;
+		} else if (unlikely(ret < 0)) { /* no choice here but crash */
+			UK_CRASH("Could not signal parent (%d)\n", ret);
+		}
+#else /* !CONFIG_LIBPOSIX_PROCESS_SIGNAL */
+		nowait = true;
+#endif /* !CONFIG_LIBPOSIX_PROCESS_SIGNAL */
+	}
+
+	/* Terminate pthreads.
+	 *
+	 * This must happen before we unblock wait(), so that threads
+	 * have already been released when we reap the process.
+	 */
+	uk_pprocess_foreach_pthread(pprocess, pt, ptn)
+		pprocess_exit_pthread(pt, PP2PT_STATE(state), exit_status);
+
+	/* ---> No ops at the posix level on behalf of the current <---
+	 * ---> thread from this point                             <---
+	 */
+	UK_ASSERT(uk_list_empty(&pprocess->threads));
+
+	/* Update process state & exit_status*/
+	if (state == POSIX_PROCESS_EXITED)
+		pprocess->exit_status = exit_status & 0xff;
+	else /* POSIX_PROCESS_KILLED */
+		pprocess->exit_status = exit_status;
+	pprocess->state = state;
+
+	/* Reparent child processes to init. The pprocess_parent
+	 * variable still refers to the original parent.
+	 */
+	pprocess_reparent_children(pprocess);
+
+	/* Notify handlers */
+	event_data.thread = NULL;
+	event_data.pid = pprocess->pid;
+	ret = uk_raise_event(POSIX_PROCESS_EXIT_EVENT, &event_data);
+	if (unlikely(ret < 0))
+		UK_CRASH("POSIX_PROCESS_EXIT_EVENT handler returned error\n");
+
+	/* Unblock wait */
+	if (parent_process && nowait == false) {
+		uk_pprocess_foreach_pthread(parent_process, pt, ptn) {
+			if (PTHREAD_WAITING_FOR_PID(pt, uk_sys_getpid()) ||
+			    PTHREAD_BLOCKING_ON_SIGNAL(pt)) {
+				uk_semaphore_up(&parent_process->wait_semaphore);
+				break;
+			}
+		}
+	} else if (nowait == true) { /* release now */
+		pprocess_release(pprocess);
```

Comment:
Can you leave a comment here that says that init process needs to manually call the release function since it cannot reach `nowait == true`?

Reply:
init is a bit special as it cannot be waited for. So it can either be released via the return path (), otherwise if it hasn't returned until shutdown, it will be released (after being force-terminated) with the system going down.